datafusion/functions/src/unicode/reverse.rs

```diff
@@ -119,12 +119,21 @@ fn reverse_impl<'a, T: OffsetSizeTrait, V: StringArrayType<'a>>(
 ) -> Result<ArrayRef> {
     let mut builder = GenericStringBuilder::<T>::with_capacity(string_array.len(), 1024);
 
-    let mut reversed = String::new();
+    let mut string_buf = String::new();
     for string in string_array.iter() {
         if let Some(s) = string {
-            reversed.extend(s.chars().rev());
-            builder.append_value(&reversed);
-            reversed.clear();
+            if s.is_ascii() {
+                // SAFETY: Since the original string was ASCII, reversing the bytes still results in valid UTF-8.
+                let reversed = unsafe {
+                    // reverse bytes directly since ASCII characters are single bytes
+                    String::from_utf8_unchecked(s.bytes().rev().collect::<Vec<u8>>())
```

Comment:
We could avoid a allocation here by copying into a `Vec<u8>` buffer
```
bytes_buf.extend(s.bytes())
bytes_buf.reverse()
...
bytes_buf.clear()
```

Reply:
Got an improvement of ~30% at most
[https://[gist.githubusercontent.com/UBarney/66aabd8f8ce6b9b7d816903660a75a4c/raw/b4d6dbe3dc545d69e2ae721326bfdf58cc70cf37/gistfile1.txt]()](https://gist.githubusercontent.com/UBarney/66aabd8f8ce6b9b7d816903660a75a4c/raw/b4d6dbe3dc545d69e2ae721326bfdf58cc70cf37/gistfile1.txt)

Reply:
Got an improvement of [~30% at most](https://gist.githubusercontent.com/UBarney/66aabd8f8ce6b9b7d816903660a75a4c/raw/b4d6dbe3dc545d69e2ae721326bfdf58cc70cf37/gistfile1.txt)

Reply:
Nice